contracts/RollupPoB.sol

```diff
@@ -0,0 +1,331 @@
+    ) external payable {
+        require(slot >= currentSlot() + minNumSlots, 'This auction is already closed');
+        doBid(slot, beneficiaryAddress, forgerAddress, withdrawAddress, msg.sender, true);
+    }
+
+    /**
+     * @dev Receive a bid from an operator
+     * Forger address, beneficiary address, withdraw address and bonus address are submitted as parameters
+     * @param slot slot for which the operator is offering
+     * @param forgerAddress controller address
+     * @param beneficiaryAddress beneficiary address
+     * @param withdrawAddress withdraw address
+     * @param bonusAddress withdraw address
+     * @param useBonus decide whether to use the bonus saved in the smart contract
+     */
+    function bidWithDifferentAddresses(
+        uint32 slot,
+        address payable beneficiaryAddress,
+        address forgerAddress,
+        address withdrawAddress,
+        address bonusAddress,
+        bool useBonus
+    ) external payable {
+        require(slot >= currentSlot() + minNumSlots, 'This auction is already closed');
+        doBid(slot, beneficiaryAddress, forgerAddress, withdrawAddress, bonusAddress, useBonus);
+    }
+
+    /**
+     * @dev distribution of the amount
+     * @param amount amount to distribute
+     * @param op operator who will receive the amount
+     */
+    function _returnBid(uint amount, uint bonus, Operator storage op) private {
+        withdrawBid[op.withdrawAddress] += amount;
+        bonusBalance[op.bonusAddress] += bonus;
+    }
+    /**
+     * @dev function to withdraw bid
+     */
+    function withdraw() external {
+        require(withdrawBid[msg.sender] > 0, 'You cannot withdraw the amount');
+        uint auxAmount = withdrawBid[msg.sender];
+        withdrawBid[msg.sender] = 0;
+        msg.sender.transfer(auxAmount);
+    }
+
+    /**
+     * @dev operator commits data and forge a batch
+     * @param compressedTx data committed by the operator. Represents off-chain transactions
+     * @param proofA zk-snark input
+     * @param proofB zk-snark input
+     * @param proofC zk-snark input
+     * @param input public zk-snark inputs
+     */
+    function commitAndForge(
+        bytes calldata compressedTx,
+        uint[2] calldata proofA,
+        uint[2][2] calldata proofB,
+        uint[2] calldata proofC,
+        uint[8] calldata input
+    ) external virtual {
+        uint32 slot = currentSlot();
+        Operator storage op = slotWinner[slot];
+        // message sender must be the controller address
+        require(msg.sender == op.forgerAddress, 'message sender must be forgerAddress');
+        // Check if deadline has been achieved to not commit any more data
+        uint blockDeadline = getBlockBySlot(slot + 1) - SLOT_DEADLINE;
+        require(getBlockNumber() < blockDeadline, 'not possible to commit data after deadline');
+        uint256 offChainHash = hashOffChainTx(compressedTx, MAX_TX);
+        emit dataCommitted(offChainHash);
```

Comment:
I will emit the event at the end of the function, just for clarity :)
  - same applies to `commitAndForgeDeadline` function